[X86] Insert CALLSEQ when lowering GlobalTLSAddress for ELF targets

ritter-x2a opened the pull request:
When lowering a TLS address for an ELF target, we introduce a call to obtain the TLS base address. So far, we do not insert CALLSEQ_START/END markers around this call when it is generated, but use a custom inserter to insert them in a later phase. This is problematic, since the TLS address call can land in a CALLSEQ for another calls before it is wrapped in its own CALLSEQ. That results in nested CALLSEQs, which are illegal and cause errors when expensive checks are enabled, e.g., in issues #45574 and #98042.

This patch instead wraps each TLS address call in a CALLSEQ when it is generated so that instruction selection can avoid nested CALLSEQs. This is an alternative to PR #106965, which instead changes the custom inserter to avoid generating CALLSEQs when the TLS address call is already in a CALLSEQ.

This patch also effectively reverts commit [228978c](https://github.com/llvm/llvm-project/commit/228978c0dcfc9a9793f3dc8a69f42471192223bc), which introduced the CustomInserter that so far added the CALLSEQ around TLSAddrs.

Fixes #45574 and #98042.

ritter-x2a committed:
[X86] Insert CALLSEQ when lowering GlobalTLSAddress for ELF targets

When lowering a TLS address for an ELF target, we introduce a call to obtain
the TLS base address. So far, we do not insert CALLSEQ_START/END markers around
this call when it is generated, but use a custom inserter to insert them in a
later phase. This is problematic, since the TLS address call can land in a
CALLSEQ for another calls before it is wrapped in its own CALLSEQ. That results
in nested CALLSEQs, which are illegal and cause errors when expensive checks
are enabled, e.g., in issues #45574 and #98042.

This patch instead wraps each TLS address call in a CALLSEQ when it is
generated so that instruction selection can avoid nested CALLSEQs. This is an
alternative to PR #106965, which instead changes the custom inserter to avoid
generating CALLSEQs when the TLS address call is already in a CALLSEQ.

This patch also effectively reverts commit 228978c, which introduced the
CustomInserter that so far added the CALLSEQ around TLSAddrs.

Fixes #45574 and #98042.

arsenm commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -18852,13 +18870,20 @@ GetTLSADDR(SelectionDAG &DAG, SDValue Chain, GlobalAddressSDNode *GA,
+  if (LoadGlobalBaseReg) {
+    SDValue InGlue;
+    Chain = DAG.getCopyToReg(Chain, dl, X86::EBX,
+                             DAG.getNode(X86ISD::GlobalBaseReg, SDLoc(), PtrVT),
```
Use dl?

arsenm commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -18852,13 +18870,20 @@ GetTLSADDR(SelectionDAG &DAG, SDValue Chain, GlobalAddressSDNode *GA,
   }
+  Chain = DAG.getCALLSEQ_END(Chain, 0, 0, Chain.getValue(1), dl);
 
   // TLSADDR will be codegen'ed as call. Inform MFI that function has calls.
```
Is this redundant now?

ritter-x2a replied:
Indeed, the call to `setAdjustsStack` is. It is set in `FinalizeISel` when the callseq markers are seen. Removed in cf38dad8a942be3762bbd61fc27920234f32395b. Thanks!
The `setHasCalls` call seems not to be redundant.

arsenm commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -18841,8 +18842,25 @@ GetTLSADDR(SelectionDAG &DAG, SDValue Chain, GlobalAddressSDNode *GA,
     // Reuse existing GetTLSADDR node if we can find it.
-    if (UI != TGA->use_end())
-      return SDValue(*UI->use_begin()->use_begin(), 0);
+    if (UI != TGA->use_end()) {
```
This is awkward parsing of users of a target constant. Can you just use getGluedUser()?

arsenm commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -18841,8 +18842,25 @@ GetTLSADDR(SelectionDAG &DAG, SDValue Chain, GlobalAddressSDNode *GA,
+             "Unexpected TLSDESC DAG");
+      // The Add generated at the final return of this function uses
+      // CopyFromReg.
+      auto AddOp = CopyFromRegOp->use_begin();
+      assert(AddOp->getOpcode() == ISD::ADD && "Unexpected TLSDESC DAG");
+      return SDValue(*AddOp, 0);
```
Assuming this will be the only use doesn't feel great. Can you just find the root node, and let the auto-CSE property of the DAG take care of this by simply trying to recreate the add at the end?

ritter-x2a committed:
fixup! [X86] Insert CALLSEQ when lowering GlobalTLSAddress for ELF targets

Use dl for GlobalBaseReg node, remove redundant setAdjustsStack call

ritter-x2a committed:
fixup! fixup! [X86] Insert CALLSEQ when lowering GlobalTLSAddress for ELF targets

Reorganize how previously generated TLS base addresses are reused

arsenm reviewed (commented):
I'm assuming this is a global / function scope constant. If you ignore the incoming chain, and just use DAG.getEntryNode, do you find the pre-existing node on later attempts at construction?

ritter-x2a commented:
> I'm assuming this is a global / function scope constant. If you ignore the incoming chain, and just use DAG.getEntryNode, do you find the pre-existing node on later attempts at construction?

If I
- replace the incoming `Chain` argument of `GetTLSADDR` with `DAG.getEntryNode()`,
- remove the logic to retrieve the previously generated node, and
- just generate the nodes (including the CALLSEQ markers, the call, the load from the segment register, and the addition),

the resulting Node of a second invocation of GetTLSADDR for a function is not the same as that of the first invocation. The call and the CALLSEQ markers (and the addition that uses the call result) are not deduplicated, even though the sequence is based on the same incoming chain.
In the generated code for such a case, two calls are then emitted instead of one.
That CALLSEQ_END markers are not CSE'd seems to be intentional, according to [a comment above getCALLSEQ_END](https://github.com/llvm/llvm-project/blob/d83485a080341688cbb9d5c07a3839bd97f9001e/llvm/include/llvm/CodeGen/SelectionDAG.h#L1095).

ritter-x2a committed:
fixup! fixup! fixup! [X86] Insert CALLSEQ when lowering GlobalTLSAddress for ELF targets

Define operand arrays inline in getNode calls. Committed this way because the github UI kept requesting refreshes when trying to commit Matt's suggested changes.

ritter-x2a commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -18828,44 +18828,67 @@ X86TargetLowering::LowerGlobalAddress(SDValue Op, SelectionDAG &DAG) const {
+      Chain = DAG.getNode(CallType, dl, NodeTys, Ops);
+    } else {
+      SDValue Ops[] = {Chain, TGA};
+      Chain = DAG.getNode(CallType, dl, NodeTys, Ops);
```
Done in 1ebbc1468822fa614e1725a2e417b0b199ff0449.

ritter-x2a commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -18828,44 +18828,67 @@ X86TargetLowering::LowerGlobalAddress(SDValue Op, SelectionDAG &DAG) const {
+                               InGlue);
+      InGlue = Chain.getValue(1);
+      SDValue Ops[] = {Chain, TGA, InGlue};
+      Chain = DAG.getNode(CallType, dl, NodeTys, Ops);
```
Done in 1ebbc1468822fa614e1725a2e417b0b199ff0449.

arsenm reviewed (approved):
Assuming it's legal, should probably use the entry node as the input chain for the sequence

ritter-x2a committed:
fixup! fixup! fixup! fixup! [X86] Insert CALLSEQ when lowering GlobalTLSAddress for ELF targets

Replace GetTLSADDR's Chain parameter with DAG.getEntryNode().

ritter-x2a commented:
> Assuming it's legal, should probably use the entry node as the input chain for the sequence

With my patches so far, every call site of `GetTLSADDR` uses `DAG.getEntryNode()` as chain argument anyway. 426eec65cea888a8b03cc2000c0aa2bb58459ce7 now removes the chain argument and uses `DAG.getEntryNode()` instead.

ritter-x2a commented:
Rebased to see if the seemingly unrelated CI fails would persist; encountered different, seemingly unrelated CI fails instead, therefore merging now.
